File: 엄마호랑이반/9주차 문제/이다영/안전 영역.java
package temp_dev;

import java.util.*;
import java.io.*;

public class Boj {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());
		
		int maxRain = 0;
		int[][] region = new int[n][n];
		for (int i = 0; i < n; i++) {
			int[] tmp = Arrays.stream(br.readLine().split(" "))
					.mapToInt(Integer::parseInt)
					.toArray();
			
			for (int j = 0; j < n; j++) {
				region[i][j] = tmp[j];
				
				maxRain = Math.max(maxRain, tmp[j]);
			} 
		}
		
		Deque<int[]> dq = new ArrayDeque<>();
		int[] dx = new int[] {-1, 0, 1, 0};
		int[] dy = new int[] {0, 1, 0, -1};
		
		int res = 0;
		for (int r = 0; r <= maxRain + 1; r++) {
			int cnt = 0; 
			boolean[][] visited = new boolean[n][n];
			
			for(int i = 0; i < n; i++) {
				for (int j = 0; j < n; j++) {
					if (region[i][j] > r && !visited[i][j]) {
						cnt++;
						visited[i][j] = true;
						dq.addLast(new int[] {i, j});
						
						while(!dq.isEmpty()) {
							int[] tmp = dq.pollFirst();
							int x = tmp[0];
							int y = tmp[1];
							
							for (int z = 0; z < 4; z++) {
								int xx = x + dx[z];
								int yy = y + dy[z];
								
								if(0 <= xx && xx < n && 0 <= yy && yy < n &&
									region[xx][yy] > r && !visited[xx][yy]) {
									visited[xx][yy] = true;
									dq.addLast(new int[] {xx, yy});
								}
							}
						}
					}
				}
			}
			
			res = Math.max(res, cnt);
		}
		
		System.out.println(res);
		
	}
}

File: 엄마호랑이반/9주차 문제/박정환/안전 영역.py
import sys

dx = [-1, 0, 1, 0]
dy = [0, 1, 0 , -1]
sys.setrecursionlimit(10**6)

def DFS(x, y, h):
    ch[x][y] = 1
    for k in range(4):
        xx = x + dx[k]
        yy = y + dy[k]
        if 0<= xx <n  and 0<=yy<n and board[xx][yy]>h and ch[xx][yy] == 0:
            DFS(xx, yy, h)

n = int(input())
board = [list(map(int, input().split())) for _ in range(n)]

res = 0
for h in range(100):
    cnt = 0
    ch = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if board[i][j] > h and ch[i][j] == 0:
                cnt += 1
                DFS(i, j, h)
    res = max(res, cnt)

print(res)
